Add com.one_ware.OneWare

HendrikMennen opened the pull request:
<!-- ⚠️  The submission PR must be against the `new-pr` branch ⚠️  -->

### Please confirm your submission meets all the criteria

<!-- Please replace each `[ ]` with `[X]` when the step is complete -->

- [X] Please describe the application briefly.
- [X] The domain used for the application ID is [controlled by the application developers][appid-domain] and the [application id guidelines][appid] are followed.
- [X] I have read the and followed all the [Submission and licence requirements][reqs].
- [X] I have [built][build] and tested the submission locally.
- [X] I am an author/developer/upstream contributor of the project.

This project is already released as a snap. I am the author. It is an IDE for electronic development and needs access to USB.  It is a free to use software but there might be some premium extensions in the future. It is my first time building a flatpak, I carefully followed the guides. The only issues I had while linting were about the mirrored screenshots/icons despite using the --mirror arguments, I guess this will be fixed with release?

The website is https://one-ware.com/studio

<!-- ⚠️  Don't modify anything below this line ⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements#control-over-domain
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

hfiguiere commented on com.one_ware.OneWare.desktop:
this should be upstream

hfiguiere commented on com.one_ware.OneWare.metainfo.xml:
this should be upstream

hfiguiere commented on com.one_ware.OneWare.metainfo.xml:
```diff
@@ -0,0 +1,48 @@
+    <name translatable="no">Hendrik Mennen</name>
+  </developer>
+  <metadata_license>CC0-1.0</metadata_license>
+  <project_license>LicenseRef-proprietary</project_license>
```
you said " It is a free software " but then this. Which is it?

HendrikMennen replied:
I think I said this wrong, I appologize. By free I wanted to say free to use, unlike many Jetbrains IDEs for example where you need to login first. The source is not entirely free (yet), which we might do at some point. A big part of the IDE system is already open source, but not everything.

hfiguiere commented on com.one_ware.OneWare.png:
this should be upstream

hfiguiere commented on flathub.json:
```diff
@@ -0,0 +1,5 @@
+{
+    "only-arches": [
+        "x86_64"
```
this is disappointing.

HendrikMennen replied:
While compiling avalonia for ARM is possible, some binaries that are needed to make this IDE work fully are not available (yet). Also I would prefer to test this before allowing arm builds

hfiguiere commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,49 @@
+      - install -D -m644 ${FLATPAK_ID}.png ${FLATPAK_DEST}/share/icons/hicolor/512x512/apps/${FLATPAK_ID}.png
+      - install -D -m644 ${FLATPAK_ID}.metainfo.xml ${FLATPAK_DEST}/share/metainfo/${FLATPAK_ID}.metainfo.xml
+    sources:
+      - type: extra-data
```
why do you use extra-data? You said you were the upstream developer. So that imply you allow redistribution (non withstanding the " It is a free software "), so extra-data should not be used here.

HendrikMennen replied:
While the IDE System I made is open source, the Studio itself (including FPGA hardware features) is not.
Thats why the sources are downloaded from our cdn.

hfiguiere replied:
note that I used the word "redistribution" here. A Free Software license would imply it, but it's not required. Since you are the upstream developer you should know whether it can be reditributed.

The use of `extra-data` is only for situation where things can be redistributed.

HendrikMennen replied:
I see, guess I missed that in the documentation. I was following an example from another flatpak that is also using Avalonia. So I guess I should move away from extra-data and instead download the sources in the build process. I will try this again later

hfiguiere commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,49 @@
+  - --share=network
+  - --socket=pulseaudio
+  - --socket=x11
+  - --filesystem=host
```
this is a very broad file system permission?

HendrikMennen replied:
I was looking at other examples for IDEs and in my opinion it does make sense to have broad file system permissions for a code editor, similar to Rider or VSCode?

HendrikMennen replied:
I changed it tom home already. This will force users to install their Quartus (or other FPGA toolchains) in home aswell. At least for quartus thats the default install location anyways so its not so bad.
![image](https://github.com/flathub/flathub/assets/25281882/a5168698-1c94-475d-b0ed-7a046925d5d0)

bbhtt commented on com.one_ware.OneWare.metainfo.xml:
```diff
@@ -0,0 +1,48 @@
+  <id>com.one_ware.OneWare</id>
+  <name>OneWare Studio</name>
+  <summary>IDE for Electronics Development</summary>
+  <developer id="one-ware.com">
```
```suggestion
  <developer id="com.one_ware">
```

bbhtt commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,49 @@
+tags:
+  - proprietary
+finish-args:
+  - --persist=.local/share/OneWareStudio
```
persist does not work here since you used host Did you see the line here? https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access

> The --persist=DIR option can be used to map directories from the user’s home directory into the sandbox filesystem. This only works if the application has no home or a broader permission like host that includes home.

HendrikMennen replied:
> persist does not work here since you used host Did you see the line here? https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access
> 
> > The --persist=DIR option can be used to map directories from the user’s home directory into the sandbox filesystem. This only works if the application has no home or a broader permission like host that includes home.

Understood, So I have two options: Remove the persist and keep it on host (like other IDEs) or try to make it work with XDG directories. Since this is a code editor that can be used for every kind of code, I prefer the first option.

bbhtt commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,49 @@
+  - proprietary
+finish-args:
+  - --persist=.local/share/OneWareStudio
+  - --device=all
```
why do you need `all` here?

HendrikMennen replied:
I need access to USB devices for programming the hardware and using components like Serial Monitor.
In the [docs](https://docs.flatpak.org/en/latest/sandbox-permissions.html#device-access) it was not clear to me if dri and input would be enough

bbhtt replied:
all is ok, input is not widely available.

HendrikMennen commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,49 @@
+  - --device=all
+  - --share=ipc
+  - --share=network
+  - --socket=pulseaudio
```
While this IDE (without extensions) doesn't use sound devices out of the box, this functionality will be probably added later with extensions to add some features regarding audio hardware development.

hfiguiere replied:
so it doesn't.
```suggestion
```

HendrikMennen commented:
Thanks for the quick review. I think I need to clarify something:

https://github.com/one-ware/OneWare is only the IDE system used by this project, it does not contain any hardware development features. Maybe I should change the name to make this more obvious.

The software available on our website or on Snap (https://snapcraft.io/oneware) uses this system as a submodule and adds Hardware Development functionality which is currently not open source.

HendrikMennen commented:
I changed the png to a svg according to the guide.
I put the resources (manifest, svg, .desktop) in a different repo to get shared with snap: 
https://github.com/one-ware/OneWareStudio.Resources

hfiguiere commented:
they should be with the upstreamm source. None of these are specific to flatpak or snap. It's part of the Xdg spec.

HendrikMennen commented:
> they should be with the upstreamm source. None of these are specific to flatpak or snap. It's part of the Xdg spec.

The upstream source is not public. And I guess its not making a lot of sense to put them into the tar.gz archives. Or am I missing something?

hfiguiere commented:
it actually does. because that's part of the distribution.

HendrikMennen commented:
it is now contained inside the tar.gz as requested

HendrikMennen commented:
Decided to open source the entire software. So I can build it inside the flatpak. Consider this on hold until I'm done with this

HendrikMennen commented:
Building the app inside the flatpak is not easily possible since there is no network available to restore nuget packages and seems to be broken the intended way.

Here is my try doing that:
https://github.com/one-ware/com.one_ware.OneWare/tree/dotnet-plugin

```
/run/build/oneware/source/studio/OneWare.Studio.Desktop/OneWare.Studio.Desktop.csproj : error NU1101: Unable to find package Microsoft.NETCore.App.Crossgen2.linux-x64. No packages exist with this id in source(s): /run/build/oneware/nuget-sources
```


Even if the generator would work, it would still be very annoying to always build these sources files.
Thats why I think its easier and more reliable to just use the prebuilt binaries.

I am happy with this now, except for the filesystem=home, since it messes up the built-in terminal aswell:
![image](https://github.com/flathub/flathub/assets/25281882/b49e3502-a24f-4e0a-be4c-ff63b19df129)

The .desktop / metainfo can be found here now:
https://github.com/one-ware/OneWare/tree/main/studio/OneWare.Studio.Desktop

Any further suggestions?

bbhtt commented:
bot, build om.one_ware.OneWare

bbhtt commented:
bot, build com.one_ware.OneWare.yml

bbhtt commented:
bot, build com.one_ware.OneWare

HendrikMennen commented:
bot, build com.one_ware.OneWare

HendrikMennen commented:
bot, build com.one_ware.OneWare

bbhtt commented:
Unless you are fixing issues, you don't have to update versions in this PR. You can do that once the repo is created.

HendrikMennen commented:
I have some memoryaccess violations with my integrated terminal that only happen inside the flatpak container. I thought I fixed them, which is why I updated here. Sadly the issue persists for now, will convert to draft until I find a solution

HendrikMennen commented:
I found a solution, which consists of not compiling with dotnet "ReadyToRun" flags. This also makes it possible to use the nuget generator and allows me to build it from source.

Is it okay to keep the nuget sources inside this repo?
I would set up a pipeline that updates them using this script https://github.com/flatpak/flatpak-builder-tools/tree/master/dotnet automatically when I release an update.

bbhtt commented:
bot, build com.one_ware.OneWare

bbhtt commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,53 @@
+sdk-extensions:
+  - org.freedesktop.Sdk.Extension.dotnet8
+build-options:
+  prepend-path: "/usr/lib/sdk/dotnet8/bin"
+  append-ld-library-path: "/usr/lib/sdk/dotnet8/lib"
+  env:
+    PKG_CONFIG_PATH: "/app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet8/lib/pkgconfig"
+  arch:
```
```suggestion
  prepend-path: "/usr/lib/sdk/dotnet8/bin"
  append-ld-library-path: "/usr/lib/sdk/dotnet8/lib"
  prepend-pkg-config-path: "/usr/lib/sdk/dotnet8/lib/pkgconfig"
  arch:
```

bbhtt commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,53 @@
+  env:
+    PKG_CONFIG_PATH: "/app/lib/pkgconfig:/app/share/pkgconfig:/usr/lib/pkgconfig:/usr/share/pkgconfig:/usr/lib/sdk/dotnet8/lib/pkgconfig"
+  arch:
+    arm:
+      env:
+        RUNTIME: "linux-arm"
```
```suggestion
```

we build for aarch64

bbhtt commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,53 @@
+      - type: git
+        url: https://github.com/one-ware/OneWare.git
+        disable-shallow-clone: true
+        branch: main
```
This needs a tag or at least a commit https://docs.flatpak.org/en/latest/module-sources.html#git-sources

HendrikMennen commented:
bot, build com.one_ware.OneWare

bbhtt commented:
Is there any part still proprietary? The license and tag in metainfo says that, I thought you said something about open sourcing it.

HendrikMennen commented:
changed it to GPL

HendrikMennen commented:
bot, build com.one_ware.OneWare

bbhtt commented on com.one_ware.OneWare.yml:
```diff
@@ -0,0 +1,49 @@
+
+command: oneware
+separate-locales: false
+tags:
+  - proprietary
```
```suggestion
```

HendrikMennen replied:
ah I missed this one, I changed it in the metainfo

bbhtt commented:
bot, build com.one_ware.OneWare
